hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/onset/OnsetChecker.java

```diff
@@ -0,0 +1,230 @@
+
+        if (!txBody.hasTransactionID()) {
+            throw new PreCheckException(INVALID_TRANSACTION_ID);
+        }
+
+        var txnId = txBody.getTransactionID();
+        if (recordCache.isReceiptPresent(txnId)) {
+            throw new PreCheckException(DUPLICATE_TRANSACTION);
+        }
+
+        if (!isPlausibleTxnFee(txBody.getTransactionFee())) {
+            throw new PreCheckException(INSUFFICIENT_TX_FEE);
+        }
+
+        if (!isPlausibleAccount(txnId.getAccountID())) {
+            throw new PreCheckException(PAYER_ACCOUNT_NOT_FOUND);
+        }
+
+        checkMemo(txBody.getMemo());
+
+        checkTimebox(txnId.getTransactionValidStart(), txBody.getTransactionValidDuration());
+    }
+
+    private static boolean isPlausibleTxnFee(final long transactionFee) {
+        return transactionFee >= 0;
+    }
+
+    private static boolean isPlausibleAccount(final AccountID accountID) {
+        return accountID.getAccountNum() > 0
+                && accountID.getRealmNum() >= 0
+                && accountID.getShardNum() >= 0;
+    }
+
+    private void checkMemo(final String memo) throws PreCheckException {
+        final var buffer = memo == null ? new byte[0] : memo.getBytes(StandardCharsets.UTF_8);
+        if (buffer.length > dynamicProperties.maxMemoUtf8Bytes()) {
+            throw new PreCheckException(MEMO_TOO_LONG);
+        }
+        for (final byte b : buffer) {
+            if (b == 0) {
+                throw new PreCheckException(INVALID_ZERO_BYTE_IN_STRING);
+            }
+        }
+    }
+
+    private void checkTimebox(final Timestamp start, final Duration duration)
+            throws PreCheckException {
+        final var validForSecs = duration.getSeconds();
+        if (validForSecs < dynamicProperties.minTxnDuration()
+                || validForSecs > dynamicProperties.maxTxnDuration()) {
+            throw new PreCheckException(INVALID_TRANSACTION_DURATION);
+        }
+
+        final var validStart = safeguardedInstant(start);
+        final var validDuration = safeguardedDuration(validForSecs, validStart);
+        final var consensusTime = Instant.now(Clock.systemUTC());
+        if (validStart.plusSeconds(validDuration).isBefore(consensusTime)) {
+            throw new PreCheckException(TRANSACTION_EXPIRED);
+        }
+        if (!validStart.isBefore(consensusTime)) {
+            throw new PreCheckException(INVALID_TRANSACTION_START);
+        }
+    }
+
+    private static Instant safeguardedInstant(final Timestamp timestamp) {
+        return Instant.ofEpochSecond(
+                Math.min(
+                        Math.max(Instant.MIN.getEpochSecond(), timestamp.getSeconds()),
+                        Instant.MAX.getEpochSecond()),
+                Math.min(
+                        Math.max(Instant.MIN.getNano(), timestamp.getNanos()),
+                        Instant.MAX.getNano()));
+    }
+
+    private long safeguardedDuration(final long validForSecs, final Instant validStart) {
```

Comment:
On each of these private methods, I think it would be good to have a comment or two, especially with some description of why the method is important (for those cases where it isn't obvious without more context). For this method for example, it wasn't obvious to me without puzzling out the implementation what the method did (ie. from the name and signature I wasn't sure what it did).